solvers/src/macro_solver/solver.rs

```diff
@@ -25,20 +25,24 @@ struct SearchNode {
     backtrack_index: u32,
 }
 
-pub struct MacroSolver {
+type ProgressCallback<'a> = dyn Fn(&[Action]) + 'a;
+pub struct MacroSolver<'a> {
     settings: Settings,
     finish_solver: FinishSolver,
     bound_solver: UpperBoundSolver,
+    progress_callback: Box<ProgressCallback<'a>>,
 }
 
-impl MacroSolver {
-    pub fn new(settings: Settings) -> MacroSolver {
+impl<'a> MacroSolver<'a> {
+    pub fn new<F>(settings: Settings, callback: F) -> MacroSolver<'a> 
```

Comment:
Remove generic parameter `F` and use `Box<ProgressCallback<'a>>` for the function argument, because:
- You are already constructing a `Box` when you create the solver in the webworker
- Each closure definition (even if the closure has the same body) is a unique type, so this will generate a new instance of `MacroSolver::new` for each test, which is probably bad for compile time.